api/v1alpha1/rabbitmqsecretenginerole_types.go

```diff
@@ -0,0 +1,176 @@
+
+	// List of topics to provide
+	// +kubebuilder:validation:Required
+	Topics []Topic `json:"topics,omitempty"`
+}
+
+type Topic struct {
+	// Name of an existing topic.
+	// +kubebuilder:validation:Required
+	TopicName string `json:"vhostName,omitempty"`
+
+	// Permissions to grant to the user in the specific vhost
+	// +kubebuilder:validation:Required
+	Permissions VhostPermissions `json:"permissions,omitempty"`
+}
+
+// Set of RabbitMQ permissions: configure, read and write.
+// By not setting a property (configure/write/read), it result in an empty string which does not match any permission.
+type VhostPermissions struct {
+	// +kubebuilder:validation:Optional
+	Configure string `json:"configure,omitempty"`
+	// +kubebuilder:validation:Optional
+	Write string `json:"write,omitempty"`
+	// +kubebuilder:validation:Optional
+	Read string `json:"read,omitempty"`
+}
+
+// RabbitMQSecretEngineRoleStatus defines the observed state of RabbitMQSecretEngineRole
+type RabbitMQSecretEngineRoleStatus struct {
+	// +patchMergeKey=type
+	// +patchStrategy=merge
+	// +listType=map
+	// +listMapKey=type
+	Conditions []metav1.Condition `json:"conditions,omitempty" patchStrategy:"merge" patchMergeKey:"type"`
+}
+
+//+kubebuilder:object:root=true
+//+kubebuilder:subresource:status
+
+// RabbitMQSecretEngineRole is the Schema for the rabbitmqsecretengineroles API
+type RabbitMQSecretEngineRole struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Spec   RabbitMQSecretEngineRoleSpec   `json:"spec,omitempty"`
+	Status RabbitMQSecretEngineRoleStatus `json:"status,omitempty"`
+}
+
+//+kubebuilder:object:root=true
+
+// RabbitMQSecretEngineRoleList contains a list of RabbitMQSecretEngineRole
+type RabbitMQSecretEngineRoleList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata,omitempty"`
+	Items           []RabbitMQSecretEngineRole `json:"items"`
+}
+
+func init() {
+	SchemeBuilder.Register(&RabbitMQSecretEngineRole{}, &RabbitMQSecretEngineRoleList{})
+}
+
+var _ vaultutils.VaultObject = &RabbitMQSecretEngineRole{}
+
+func (rabbitMQ *RabbitMQSecretEngineRole) GetPath() string {
+	return string(rabbitMQ.Spec.Path) + "/" + "roles" + "/" + rabbitMQ.Name
+}
+func (rabbitMQ *RabbitMQSecretEngineRole) GetPayload() map[string]interface{} {
+	return rabbitMQ.Spec.rabbitMQToMap()
+}
+func (rabbitMQ *RabbitMQSecretEngineRole) IsEquivalentToDesiredState(payload map[string]interface{}) bool {
+	desiredState := rabbitMQ.Spec.RMQSERole.rabbitMQToMap()
+	return reflect.DeepEqual(desiredState, payload)
+}
+
+func (rabbitMQ *RabbitMQSecretEngineRole) IsInitialized() bool {
+	return true
+}
+
+func (rabbitMQ *RabbitMQSecretEngineRole) PrepareInternalValues(context context.Context, object client.Object) error {
+	return nil
+}
+
+func (rabbitMQ *RabbitMQSecretEngineRole) IsValid() (bool, error) {
+	return true, nil
+}
+
+func convertToJson(vhosts interface{}) (string) {
```

Comment:
for this to work the json names used in the CRD must correspond to what Vault expects. This was usually not the case for me as CRD use camelcase and vault uses underscores... I had to define a xxxToMap for each nested struct.

Reply:
Updated json convertion